testing = r"""
Writing output to file: modelgenerator0.out

****Welcome to ModelGenerator v0.85****

If you use ModelGenerator in your analysis, please cite:

Thomas M Keane, Christopher J Creevey , Melissa M Pentony, Thomas J Naughton and James O McInerney (2006) Assessment of methods for amino acid matrix selection and their use on empirical data shows that ad hoc assumptions for choice of matrix are not justified, BMC Evolutionary Biology, 6:29

http://www.biomedcentral.com/1471-2148/6/29

Input sequence file: /Users/brett/tmp/output/temp.fasta
Format is FASTA
Datatype is NUCLEOTIDE
Building initial NJ tree......
Created models
Starting to optimise model parameters.....
Number of observed invariant sites: 22
Optimising model 1 of 56: JC
Optimising model 2 of 56: JC+G
Optimising model 3 of 56: JC+I
Optimising model 4 of 56: JC+I+G
Optimising model 5 of 56: F81
Optimising model 6 of 56: F81+G
Optimising model 7 of 56: F81+I
Optimising model 8 of 56: F81+I+G
Optimising model 9 of 56: HKY
Optimising model 10 of 56: HKY+G
Optimising model 11 of 56: HKY+I
Optimising model 12 of 56: HKY+I+G
Optimising model 13 of 56: K80
Optimising model 14 of 56: K80+G
Optimising model 15 of 56: K80+I
Optimising model 16 of 56: K80+I+G
Optimising model 17 of 56: SYM
Optimising model 18 of 56: SYM+G
Optimising model 19 of 56: SYM+I
Optimising model 20 of 56: SYM+I+G
Optimising model 21 of 56: GTR
Optimising model 22 of 56: GTR+G
Optimising model 23 of 56: GTR+I
Optimising model 24 of 56: GTR+I+G
Optimising model 25 of 56: TrN
Optimising model 26 of 56: TrN+G
Optimising model 27 of 56: TrN+I
Optimising model 28 of 56: TrN+I+G
Optimising model 29 of 56: TrNef
Optimising model 30 of 56: TrNef+G
Optimising model 31 of 56: TrNef+I
Optimising model 32 of 56: TrNef+I+G
Optimising model 33 of 56: TVM
Optimising model 34 of 56: TVM+G
Optimising model 35 of 56: TVM+I
Optimising model 36 of 56: TVM+I+G
Optimising model 37 of 56: TVMef
Optimising model 38 of 56: TVMef+G
Optimising model 39 of 56: TVMef+I
Optimising model 40 of 56: TVMef+I+G
Optimising model 41 of 56: TIM
Optimising model 42 of 56: TIM+G
Optimising model 43 of 56: TIM+I
Optimising model 44 of 56: TIM+I+G
Optimising model 45 of 56: TIMef
Optimising model 46 of 56: TIMef+G
Optimising model 47 of 56: TIMef+I
Optimising model 48 of 56: TIMef+I+G
Optimising model 49 of 56: K81uf
Optimising model 50 of 56: K81uf+G
Optimising model 51 of 56: K81uf+I
Optimising model 52 of 56: K81uf+I+G
Optimising model 53 of 56: K81
Optimising model 54 of 56: K81+G
Optimising model 55 of 56: K81+I
Optimising model 56 of 56: K81+I+G

		Uniform			G			I			I+G
JC		-116.05500320284096	-115.74553585410278	-115.63276616286927	-115.63659127943907	
F81		-114.0285175203941	-113.50229840420106	-113.47491048798501	-113.50395058752977	
HKY		-110.52589424495196	-109.94735991063322	-109.91386186630378	-109.96173474696772	
K80		-111.90735831014335	-111.42989164467532	-111.37133305547381	-111.57038055196625	
SYM		-111.25075588832513	-110.70665138279483	-110.77867736732837	-110.7711065438407	
GTR		-109.61214455653183	-108.5143600895575	-108.89546639810145	-108.49103608496952	
TrN		-110.47368820872894	-109.88439062975996	-109.8665893566199	-109.90103045804028	
TrNef		-111.82403331304756	-111.36462409540111	-111.27927402543556	-111.57316282805589	
TVM		-109.57145608938241	-108.60747449569169	-109.03344972490551	-108.60158496996644	
TVMef		-111.90254708615849	-111.40079866456745	-111.7028888031442	-111.30684901209227	
TIM		-110.32726741092353	-109.74263389914412	-109.85218779650488	-109.80049016194741	
TIMef		-111.61830864847987	-111.20155299543146	-118.54963585086455	-111.2530084520981	
K81uf		-110.37053527680376	-109.78773724778212	-109.8604265798504	-109.83650583203159	
K81		-111.69639369704582	-111.25759858890204	-118.68333142883817	-111.58277062972851	

****Hierarchical Likelihood Ratio Tests (hLRTs)****

Equal base frequencies
	Null model = JC
	Alternative model = F81
	2(lnL1-lnL0) = 4.052971364893722	df = 3
Using Chi-Squared Test
	P-value = 0.04387218611348909

Ti=Tv
	Null model = JC
	Alternative model = K80
	2(lnL1-lnL0) = 8.295289785395227	df = 1
Using Chi-Squared Test
	P-value < .0001

1 or 2 transition rates
	Null model = K80
	Alternative model = TrNef
	2(lnL1-lnL0) = 0.16664999419157311	df = 1
Using Chi-Squared Test
	P-value = 0.5637223660585795

1 or 2 transversion rates
	Null model = K80
	Alternative model = K81
	2(lnL1-lnL0) = 0.42192922619506135	df = 1
Using Chi-Squared Test
	P-value = 0.3582955782821934

Equals rates among sites
	Null model = K80
	Alternative model = K80+G
	2(lnL1-lnL0) = 0.9549333309360577	df = 1
Using mixed Chi-Squared Test
	P-value = 0.08348903547608266

No invariable sites
	Null model = K80
	Alternative model = K80+I
	2(lnL1-lnL0) = 1.0720505093390784	df = 1
Using mixed Chi-Squared Test
	P-value = 0.0715594261686548

Model Selected: K80
-lnL: 111.90735831014335
Simple Substitution Model:
Model of substitution: K80
Transition/transversion rate ratio kappa: 6.45

Nucleotide frequencies:
pi(A) = 0.25000
pi(C) = 0.25000
pi(G) = 0.25000
pi(T) = 0.25000

Expected transition/transversion ratio: 3.23
Expected pyrimidine transition/purine transition ratio: 1.00


****Corrected Akaike Information Criterion (AICc)****

Model Selected: K80
-lnL = 111.90735831014335
k = 6 (Branch lengths included as model parameters)
AICc = 225.8147166202867

Simple Substitution Model:
Model of substitution: K80
Transition/transversion rate ratio kappa: 6.45

Nucleotide frequencies:
pi(A) = 0.25000
pi(C) = 0.25000
pi(G) = 0.25000
pi(T) = 0.25000

Expected transition/transversion ratio: 3.23
Expected pyrimidine transition/purine transition ratio: 1.00


****Akaike Information Criterion 2 (AIC2)****

Model Selected: K80
-lnL = 111.90735831014335
k = 6 (Branch lengths included as model parameters)
AIC2 = 253.8147166202867

Simple Substitution Model:
Model of substitution: K80
Transition/transversion rate ratio kappa: 6.45

Nucleotide frequencies:
pi(A) = 0.25000
pi(C) = 0.25000
pi(G) = 0.25000
pi(T) = 0.25000

Expected transition/transversion ratio: 3.23
Expected pyrimidine transition/purine transition ratio: 1.00


****Bayesian Information Criterion (BIC)****

Model Selected: K80
-lnL = 111.90735831014335
k = 6 (Branch lengths included as model parameters)
BIC = 245.48022409615206

Simple Substitution Model:
Model of substitution: K80
Transition/transversion rate ratio kappa: 6.45

Nucleotide frequencies:
pi(A) = 0.25000
pi(C) = 0.25000
pi(G) = 0.25000
pi(T) = 0.25000

Expected transition/transversion ratio: 3.23
Expected pyrimidine transition/purine transition ratio: 1.00


Model-fit Ranking according to all measures
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
	AIC1			Model		Ln			AIC2			Model		Ln			BIC			Model		Ln
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

0	235.8147166202867	K80		-111.90735831014335	253.8147166202867	K80		-111.90735831014335	245.48022409615206	K80		-111.90735831014335	
1	236.74266611094762	K80+I		-111.37133305547381	257.11000640568193	JC		-116.05500320284096	248.0190914994572	K80+I		-111.37133305547381	
2	236.85978328935064	K80+G		-111.42989164467532	257.11000640568193	JC		-116.05500320284096	248.13620867786022	K80+G		-111.42989164467532	
3	237.39278739409164	K81		-111.69639369704582	257.74266611094765	K80+I		-111.37133305547381	248.66921278260122	K81		-111.69639369704582	
4	237.64806662609513	TrNef		-111.82403331304756	257.8597832893506	K80+G		-111.42989164467532	248.9244920146047	TrNef		-111.82403331304756	
5	238.51519717780408	K81+G		-111.25759858890204	258.39278739409167	K81		-111.69639369704582	250.16459596890306	JC		-116.05500320284096	
6	238.5585480508711	TrNef+I		-111.27927402543556	258.6480666260951	TrNef		-111.82403331304756	250.16459596890306	JC		-116.05500320284096	
7	238.72924819080222	TrNef+G		-111.36462409540111	261.26553232573855	JC+I		-115.63276616286927	251.4025404789579	K81+G		-111.25759858890204	
8	239.0517884899039	HKY		-110.52589424495196	261.4910717082056	JC+G		-115.74553585410278	251.44589135202492	TrNef+I		-111.27927402543556	
9	239.1407611039325	K80+I+G		-111.57038055196625	262.5151971778041	K81+G		-111.25759858890204	251.61659149195603	TrNef+G		-111.36462409540111	
10	239.23661729695974	TIMef		-111.61830864847987	262.5585480508711	TrNef+I		-111.27927402543556	252.0281044050863	K80+I+G		-111.57038055196625	
11	239.82772373260755	HKY+I		-109.91386186630378	262.7292481908022	TrNef+G		-111.36462409540111	252.12396059811354	TIMef		-111.61830864847987	
12	239.89471982126645	HKY+G		-109.94735991063322	263.1407611039325	K80+I+G		-111.57038055196625	252.9310398016039	JC+I		-115.63276616286927	
13	240.4031059908629	TIMef+G		-111.20155299543146	263.2366172969597	TIMef		-111.61830864847987	253.15657918407092	JC+G		-115.74553585410278	
14	240.74107055360753	K81uf		-110.37053527680376	266.0517884899039	HKY		-110.52589424495196	253.55004970370192	HKY		-110.52589424495196	
15	240.9473764174579	TrN		-110.47368820872894	266.27318255887815	JC+I+G		-115.63659127943907	254.90136720466091	TIMef+G		-111.20155299543146	
16	241.14632565611177	TrNef+I+G	-111.57316282805589	267.4031059908629	TIMef+G		-111.20155299543146	255.6445868699098	TrNef+I+G	-111.57316282805589	
17	241.16554125945703	K81+I+G		-111.58277062972851	268.0570350407882	F81		-114.0285175203941	255.66380247325503	K81+I+G		-111.58277062972851	
18	241.57547449556424	K81uf+G		-109.78773724778212	268.1463256561118	TrNef+I+G	-111.57316282805589	255.93690285904978	HKY+I		-109.91386186630378	
19	241.7208531597008	K81uf+I		-109.8604265798504	268.16554125945703	K81+I+G		-111.58277062972851	256.0038989477087	HKY+G		-109.94735991063322	
20	241.7331787132398	TrN+I		-109.8665893566199	268.805094172317	TVMef		-111.90254708615849	256.303355386115	TVMef		-111.90254708615849	
21	241.76878125951993	TrN+G		-109.88439062975996	269.8277237326075	HKY+I		-109.91386186630378	256.5496079473877	JC+I+G		-115.63659127943907	
22	241.80509417231698	TVMef		-111.90254708615849	269.89471982126645	HKY+G		-109.94735991063322	256.8502496800498	K81uf		-110.37053527680376	
23	241.92346949393544	HKY+I+G		-109.96173474696772	270.7410705536075	K81uf		-110.37053527680376	256.944378341942	F81		-114.0285175203941	
24	242.11000640568193	JC		-116.05500320284096	270.9473764174579	TrN		-110.47368820872894	257.0565555439001	TrN		-110.47368820872894	
25	242.11000640568193	JC		-116.05500320284096	271.94982097597006	F81+I		-113.47491048798501	258.6106909030925	SYM		-111.25075588832513	
26	242.50151177665026	SYM		-111.25075588832513	272.0045968084021	F81+G		-113.50229840420106	258.61519603063846	TIMef+I+G	-111.2530084520981	
27	242.5060169041962	TIMef+I+G	-111.2530084520981	272.50151177665026	SYM		-111.25075588832513	258.91077645557715	TVMef+G		-111.40079866456745	
28	242.65453482184705	TIM		-110.32726741092353	272.5060169041962	TIMef+I+G	-111.2530084520981	259.2955715346507	K81uf+G		-109.78773724778212	
29	242.8015973291349	TVMef+G		-111.40079866456745	272.8015973291349	TVMef+G		-111.40079866456745	259.4409501987873	K81uf+I		-109.8604265798504	
30	243.14291217876482	TVM		-109.57145608938241	273.40577760628844	TVMef+I		-111.7028888031442	259.44808218976806	F81+I		-113.47491048798501	
31	243.21494899138338	TVM+G		-108.60747449569169	274.5754744955642	K81uf+G		-109.78773724778212	259.45327575232625	TrN+I		-109.8665893566199	
32	243.26553232573855	JC+I		-115.63276616286927	274.7208531597008	K81uf+I		-109.8604265798504	259.4888782986064	TrN+G		-109.88439062975996	
33	243.4057776062884	TVMef+I		-111.7028888031442	274.7331787132398	TrN+I		-109.8665893566199	259.50285802220014	F81+G		-113.50229840420106	
34	243.41330276558966	SYM+G		-110.70665138279483	274.7687812595199	TrN+G		-109.88439062975996	259.51495673273064	TVMef+I		-111.7028888031442	
35	243.48526779828825	TIM+G		-109.74263389914412	274.92346949393544	HKY+I+G		-109.96173474696772	259.6435665330219	HKY+I+G		-109.96173474696772	
36	243.49107170820557	JC+G		-115.74553585410278	275.65453482184705	TIM		-110.32726741092353	260.3746318609335	TIM		-110.32726741092353	
37	243.55735473465674	SYM+I		-110.77867736732837	276.41330276558966	SYM+G		-110.70665138279483	261.1333998046761	SYM+G		-110.70665138279483	
38	243.67301166406318	K81uf+I+G	-109.83650583203159	276.55735473465677	SYM+I		-110.77867736732837	261.2774517737432	SYM+I		-110.77867736732837	
39	243.70437559300976	TIM+I		-109.85218779650488	277.00790117505954	F81+I+G		-113.50395058752977	262.333795063271	TVMef+I+G	-111.30684901209227	
40	243.80206091608056	TrN+I+G		-109.90103045804028	277.3666628576764	K81+I		-118.68333142883817	262.4739271304955	TVM		-109.57145608938241	
41	244.0570350407882	F81		-114.0285175203941	277.61369802418454	TVMef+I+G	-111.30684901209227	262.81628275001896	TIM+G		-109.74263389914412	
42	244.06689944981102	TVM+I		-109.03344972490551	279.14291217876485	TVM		-109.57145608938241	263.00402661579386	K81uf+I+G	-109.83650583203159	
43	244.61369802418454	TVMef+I+G	-111.30684901209227	279.48526779828825	TIM+G		-109.74263389914412	263.03539054474044	TIM+I		-109.85218779650488	
44	244.94982097597003	F81+I		-113.47491048798501	279.6730116640632	K81uf+I+G	-109.83650583203159	263.1170803015018	F81+I+G		-113.50395058752977	
45	245.0045968084021	F81+G		-113.50229840420106	279.7043755930098	TIM+I		-109.85218779650488	263.13307586781127	TrN+I+G		-109.90103045804028	
46	245.028720179115	GTR+G		-108.5143600895575	279.80206091608056	TrN+I+G		-109.90103045804028	264.1568818557583	TVM+G		-108.60747449569169	
47	245.20316993993288	TVM+I+G		-108.60158496996644	281.5422130876814	SYM+I+G		-110.7711065438407	264.8732280394121	SYM+I+G		-110.7711065438407	
48	245.22428911306366	GTR		-109.61214455653183	282.0992717017291	TIMef+I		-118.54963585086455	265.00883231418595	TVM+I		-109.03344972490551	
49	245.27318255887815	JC+I+G		-115.63659127943907	282.2149489913834	TVM+G		-108.60747449569169	266.16622197743857	GTR		-109.61214455653183	
50	245.5422130876814	SYM+I+G		-110.7711065438407	283.066899449811	TVM+I		-109.03344972490551	266.2540061588301	K81+I		-118.68333142883817	
51	245.60098032389482	TIM+I+G		-109.80049016194741	284.22428911306366	GTR		-109.61214455653183	266.5429131882697	TIM+I+G		-109.80049016194741	
52	245.7909327962029	GTR+I		-108.89546639810145	284.6009803238948	TIM+I+G		-109.80049016194741	267.58157095613416	GTR+G		-108.5143600895575	
53	246.98207216993904	GTR+I+G		-108.49103608496952	287.028720179115	GTR+G		-108.5143600895575	267.75602071695204	TVM+I+G		-108.60158496996644	
54	247.00790117505954	F81+I+G		-113.50395058752977	287.2031699399329	TVM+I+G		-108.60158496996644	268.343783573222	GTR+I		-108.89546639810145	
55	253.36666285767635	K81+I		-118.68333142883817	287.7909327962029	GTR+I		-108.89546639810145	269.5975329155271	TIMef+I		-118.54963585086455	
56	255.0992717017291	TIMef+I		-118.54963585086455	291.98207216993904	GTR+I+G		-108.49103608496952	271.1458408596024	GTR+I+G		-108.49103608496952	

Analysis took: 0 minutes 5 seconds
"""

from pyparsing import (
    Word, Dict, OneOrMore, alphas, nums, Suppress, Optional, Group, stringEnd,
    delimitedList, pythonStyleComment, ParseException, line, lineno, col,
    Keyword, ParserElement, ParseException, LineStart, SkipTo, LineEnd,
    )

TABLEHEADER = LineStart() + Word("-") + LineEnd()
INTEGER = Word(nums)
FLOAT = Word(nums + '.-')
MODEL = Word(alphas + nums + '+')

RGROUP = Group(FLOAT + MODEL + FLOAT)

row = (LineStart() + INTEGER + RGROUP + RGROUP + RGROUP + Suppress(LineEnd()))

timing = Group(Suppress("Analysis took:") + INTEGER + Suppress("minutes") + INTEGER
               + Suppress("seconds"))

       # + FLOAT + FLOAT + MODEL
    # + FLOAT + LineEnd())
outparser = (Suppress(SkipTo(TABLEHEADER) + TABLEHEADER + SkipTo(TABLEHEADER) +
                     TABLEHEADER) + OneOrMore(row) + timing + stringEnd)
res = outparser.parseString(testing)
print res





